ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -0,0 +1,291 @@
+use crate::ff::{Field, Invert};
+
+/// function for generating canonical evaluation points
+/// overwrites `evaluation_points`
+pub fn generate_evaluation_points<F>(evaluation_points: &mut [F]) -> ()
+where
+    F: Field,
+{
+    // check that field is large enough to hold evaluation points
+    debug_assert!(F::BITS > usize::BITS - evaluation_points.len().leading_zeros() + 1);
+    // compute evaluation points `0`, `1`, `2`, ...
+    evaluation_points
+        .iter_mut()
+        .enumerate()
+        .for_each(|(i, p)| *p = F::try_from(i as u128).unwrap());
+}
+
+/// computes the lagrange denominators for the base polynomials
+/// the denominators change for different evaluation points
+/// the function multiplies the denominators to `denominators`, i.e. input `F::ONE`
+pub fn compute_lagrange_denominator<F>(evaluation_points: &[F], denominators: &mut [F]) -> ()
+where
+    F: Field + Invert,
+{
+    debug_assert_eq!(evaluation_points.len(), denominators.len());
+    // check that differences are non-zero
+    #[cfg(debug_assertions)]
+    {
+        for i in 0..evaluation_points.len() {
+            for j in 0..evaluation_points.len() {
+                debug_assert_eq!(
+                    (
+                        i,
+                        j,
+                        evaluation_points[i] - evaluation_points[j] != F::ZERO || j == i
+                    ),
+                    (i, j, true)
+                )
+            }
+        }
+    }
+    // compute denominators:
+    for i in 0..evaluation_points.len() {
+        if i > 0 {
+            for j in 0..i {
+                denominators[i] *= evaluation_points[i] - evaluation_points[j];
+            }
+        }
+        if i + 1 < evaluation_points.len() {
+            for j in i + 1..evaluation_points.len() {
+                denominators[i] *= evaluation_points[i] - evaluation_points[j];
+            }
+        }
+        denominators[i] = denominators[i].invert();
+    }
+}
+
+/// computes the lagrange base polynomials
```

Comment:
I don't understand the nomenclature here, can you be a bit more specific what we are trying to achieve?